tox_docker/tox4/plugin.py

```diff
@@ -0,0 +1,122 @@
+from typing import Dict, List
+
+from tox.config.cli.parser import ToxParser
+from tox.config.main import Config
+from tox.config.sets import ConfigSet
+from tox.config.types import EnvList
+from tox.execute.api import Outcome
+from tox.plugin import impl
+from tox.tox_env.api import ToxEnv
+
+from tox_docker.config import ContainerConfig, RunningContainers
+from tox_docker.plugin import (
+    docker_health_check,
+    docker_pull,
+    docker_run,
+    get_env_vars,
+    HealthCheckFailed,
+    stop_containers,
+)
+from tox_docker.tox4.config import (
+    discover_container_configs,
+    EnvRunningContainers,
+    parse_container_config,
+)
+from tox_docker.tox4.log import log
+
+CONTAINER_CONFIGS: Dict[str, ContainerConfig] = {}
+ENV_CONTAINERS: EnvRunningContainers = {}
+
+
+@impl
+def tox_add_core_config(core_conf: ConfigSet, config: Config) -> None:
+    container_config_names = discover_container_configs(config)
+
+    # validate command line options
+    for container_name in config.options.docker_dont_stop:
+        if container_name not in container_config_names:
+            raise ValueError(
+                f"Container {container_name!r} not found (from --docker-dont-stop)"
+            )
+
+    for container_name in container_config_names:
+        CONTAINER_CONFIGS[container_name] = parse_container_config(
+            config, container_name, container_config_names
+        )
+
+
+@impl
+def tox_before_run_commands(tox_env: ToxEnv) -> None:
+    tox_env.conf.add_config(
+        keys=["docker"],
+        of_type=EnvList,
+        default=EnvList([]),
+        desc="docker image configs to load",
+        post_process=list,  # type: ignore
+    )
+
+    container_names = tox_env.conf["docker"]
+    env_container_configs = []
+
+    seen = set()
+    for container_name in container_names:
+        if container_name not in CONTAINER_CONFIGS:
+            raise ValueError(f"Missing [docker:{container_name}] in tox.ini")
+        if container_name in seen:
+            raise ValueError(f"Container {container_name!r} specified more than once")
+        seen.add(container_name)
+        env_container_configs.append(CONTAINER_CONFIGS[container_name])
+
+    for container_config in env_container_configs:
+        docker_pull(container_config, log)
```

Comment:
I feel these should be on the docker config set, so they're implicitly shared. If two enmv requires the same image you don't need to force pull twice.

Reply:
It won't pull if it's already local -- just as the `docker pull` CLI command won't.

This can't be moved to the config set, it needs to support tox3. The things in `tox_docker/` but not in the tox3 or tox4 sub-packages are "generic" code to support both versions.

Reply:
Why do you want to support both tox 3 and 4 going ahead? tox itself will not support version 3... so I'm not sure I see the point :shrug:

Reply:
I expect many folks will not upgrade immediately, so I'm going to support both versions for some time.

Reply:
Why? Ignoring plugins (and some features that are deprecated for years) tox 4 is a drop-in replacement for tox 3. Even if wouldn't be I'd not expect you to keep adding improvements for both.